### ISSUE TITLE
filename is in correct when we create multiple instance of same file?[Android 9]

### ISSUE THREAD

**Issue by dhaval-k-simformsolutions** [alexrintt/shared-storage#22]
**Describe the bug**
The filenames are wrong when we save the same file multiple times in Android 9 version phones.

**To Reproduce**
1. Grant document tree permission to any directory. 
2. Save the file using `createFileAsString` method in the storage
3. save the Same file multiple times. like "myfile.csv"

**Expected behavior**
When we save the same file multiple times it should create the files as below
First time - `myfile.csv`
Second time - `myfile (1).csv`
Thrid time - `myfile (2).csv`

**Note:** It's working fine in Android 10 and above versions, But I have check-in 2 phones(Vivo and Asus) with android 9 and in those phones, it generates the above files as below which is a totally incorrect file format.

First time - `myfile.csv`
Second time - `myfile.csv (1)`
Thrid time - `myfile.csv (2)`

**Screenshots**
![Screenshot_20220413-142511](https://user-images.githubusercontent.com/90894202/163184647-78ca15ce-b973-4fe2-9826-f05d7e20f656.png)

**Comment by lakscastro** [comment]
Are you using the mimeType field when creating the file with `createFileAsString`?

**Comment by dhaval-k-simformsolutions** [comment]
No, I am passing the mime type as well while `createFileAsString`. Please find the code below for the same.

```
      final newDocumentFile = await createFileAsString(
        selectedUriDir,
        mimeType: 'text/csv',
        content: csvData,
        displayName: fileName,
      );
```

**Comment by lakscastro** [comment]
I see, then I need to search further through Android 9 and above. Thanks for the report

**Comment by dhaval-k-simformsolutions** [comment]
Yeah sure

**Comment by dhaval-k-simformsolutions** [comment]
@lakscastro  As a temporary workaround of I am able to find the existing file with the same name using the `findFile` method and I just want to simply delete the older file and want to create the new one so every time there will be only one file will be there with the unique file name. 

But the problem in the above workaround is to delete the file I required file path using which but there is no way using which I can get the file path of that file or If you provide method using which I can delete that file on persisted URI then it will work me as well. 

Still, if there is a way to get the file path of persisted URI then it will be great. There is some method available in the native on StackOverflow as `getRealPathFromUri` I tried the same but got the below exception with that method.

`getRealPathFromURI Exception : java.lang.UnsupportedOperationException: Unsupported Uri content://com.android.externalstorage.documents/tree/primary%3ADirectorName`

**Comment by lakscastro** [comment]
> is to delete the file I required file path

This is no longer available for recent Android APIs so this will not be implemented in this module (SAF)

But you can still use the `delete` method to delete a  given `Uri`. Usage:

```dart
delete(uri); /// uri should be a Uri that you have write access
```

**Comment by dhaval-k-simformsolutions** [comment]
> > is to delete the file I required file path
> 
> This is no longer available for recent Android APIs so this will not be implemented in this module (SAF)
> 
> But you can still use the `delete` method to delete a given `Uri`. Usage:
> 
> ```dart
> delete(uri); /// uri should be a Uri that you have write access
> ```

delete method will delete the complete URI I guess means it will delete the complete directory my requirement is I want to just delete the file not directory

**Comment by lakscastro** [comment]
This is what you can do, take a look at `example/` folder, we have 2 kind of Uris: `Document` (File) and `Tree` (File) Uris, you want to delete the Document Uri, the thing I don't know how to do yet is how you can do it directly to you target `DocumentUri`

**Comment by lakscastro** [comment]
@dhaval-k-simformsolutions 

Do you remember what data was here?

```dart
      final newDocumentFile = await createFileAsString(
        selectedUriDir,
        mimeType: 'text/csv',
        content: csvData,
        displayName: fileName, // here
      );
```

Do you remember if you were sending the `displayName` arg with the file extension? Like `myfilename.png` or just `myfilename`?

**Comment by dhaval-k-simformsolutions** [comment]
> @dhaval-k-simformsolutions
> 
> Do you remember what data was here?
> 
> ```dart
>       final newDocumentFile = await createFileAsString(
>         selectedUriDir,
>         mimeType: 'text/csv',
>         content: csvData,
>         displayName: fileName, // here
>       );
> ```
> 
> Do you remember if you were sending the `displayName` arg with the file extension? Like `myfilename.png` or just `myfilename`?

Hi @lakscastro  
I was passing the filename along with the file extension.

**Comment by lakscastro** [comment]
Hey @dhaval-k-simformsolutions 

Sorry for bothering you again, but can you please share some specs of your device? I'm not able to reproduce this bug neither on Android 11 (Xiaomi), Android 10 (Samsung) or Android 7.1 (Pixel)

Some useful info would be:

- Device Android Version
- Device Model (LG? Samsung? Xiaomi? Pixel? etc.)
- Device File Explorer; Are you using the default one provided by your OS or do you installed a third party file explorer?

**Comment by lakscastro** [comment]
After a while I noticed this happens only to a few kind of mime types; and the `csv` is one of those. The fix is to use the complete mime type, in your case `text/comma-separated-values`

```dart
documentFile?.createFileAsString(
  mimeType: 'text/comma-separated-values', // Don't work with `text/csv`
  content: 'Sample File Content',
  displayName: 'Sample File',
);
```

Also, we need avoid sending the extension along with the `displayName`, so is recommended to use `myfilename` instead `myfilename.png` for example

**Comment by dhaval-k-simformsolutions** [comment]
> **Note:** It's working fine in Android 10 and above versions, But I have check-in 2 phones(Vivo and Asus) with android 9 and in those phones, it generates the above files as below which is a totally incorrect file format.

@lakscastro 

As I have already mentioned the device information in the issue, still mentioning the same again below.
**Note:** It's working fine in Android 10 and above versions, But I have check-in **2 phones(Vivo and Asus) with android 9** and in those phones, it generates the above files as below which is the totally incorrect file format.

**Comment by dhaval-k-simformsolutions** [comment]
> After a while I noticed this happens only to a few kind of mime types; and the `csv` is one of those. The fix is to use the complete mime type, in your case `text/comma-separated-values`
> 
> ```dart
> documentFile?.createFileAsString(
>   mimeType: 'text/comma-separated-values', // Don't work with `text/csv`
>   content: 'Sample File Content',
>   displayName: 'Sample File',
> );
> ```
> 
> Also, we need avoid sending the extension along with the `displayName`, so is recommended to use `myfilename` instead `myfilename.png` for example

Okay, I will try the same without extension with filename while creating the file and will let you know the updates.

**Comment by dhaval-k-simformsolutions** [comment]
> > After a while I noticed this happens only to a few kind of mime types; and the `csv` is one of those. The fix is to use the complete mime type, in your case `text/comma-separated-values`
> > ```dart
> > documentFile?.createFileAsString(
> >   mimeType: 'text/comma-separated-values', // Don't work with `text/csv`
> >   content: 'Sample File Content',
> >   displayName: 'Sample File',
> > );
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also, we need avoid sending the extension along with the `displayName`, so is recommended to use `myfilename` instead `myfilename.png` for example
> 
> Okay, I will try the same without extension with filename while creating the file and will let you know the updates.

I have tried to createFileAsString with the only file name(without extension), and it creates a file as shown below which on openable.
![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/90894202/165028574-bfd8117f-e796-47c0-af05-d2db1435937d.png)

**Comment by dhaval-k-simformsolutions** [comment]
@lakscastro Right now where I get the package from the master branch it gives me the below error while running the project.
```
e: /Users/manoj.padiya/.pub-cache/git/shared-storage-2ed784c56f885856e559378699df02736e9ee283/android/src/main/kotlin/io/lakscastro/sharedstorage/storageaccessframework/DocumentFileApi.kt: (10, 8): Unresolved reference: com
e: /Users/manoj.padiya/.pub-cache/git/shared-storage-2ed784c56f885856e559378699df02736e9ee283/android/src/main/kotlin/io/lakscastro/sharedstorage/storageaccessframework/DocumentFileApi.kt: (209, 41): Unresolved reference: child
```
Can you please check into it.

**Comment by lakscastro** [comment]
Merged #45 which fix this.

**Comment by alexrintt** [comment]
Closing as inactive.
